Fix the _array_ methods to avoid deprecation warnings

SamLubelsky opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed

Fixes #26302

I finished the work started in PR #26303 by adding test cases for the __array__ methods.

Note: currently the tests with copy=False are commented out because there is an issue 
with numpy passing the copy argument into the __array__ function.  As a result, __array__ 
will not raise an error as it should, so this test must be commented out for now.

see this comment for confirmation:
 https://github.com/numpy/numpy/issues/25941#issuecomment-2010343170.
#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

RajSapale04 committed:
Changed the signature methods for _array_ methods in Sympy to :

   def __array__(self, dtype=object, copy=None):
        if copy is not None and not copy:
            raise TypeError("Cannot implement copy=False when converting Matrix to ndarray")

RajSapale04 committed:
author: add Raj Sapale to .mailmap

SamLubelsky committed:
Added test cases for the new __array__ methods(Copy=true/false/None)

Currently copy=False is commented out due to a bug
in numpy. see:
https://github.com/numpy/numpy/issues/25941#issuecomment-2010343170

SamLubelsky committed:
fixed trailing whitespace

SamLubelsky committed:
actually fixed trailing whitespace

SamLubelsky committed:
now imports skip properly.

oscarbenjamin commented on sympy/matrices/tests/test_matrixbase.py:
```diff
@@ -351,6 +351,16 @@ def test_replace_map():
+    np_array = array([[1,2], [3,4]])
+    assert array_equal(array(A), np_array)
+    assert array_equal(array(A, copy=True), np_array)
+    #raises(TypeError, lambda: array(A, copy=False)) TODO: Uncomment this whenever copy variable properly passes to __array__
```
This works as intended if numpy 2.0.0rc1 is installed.

We should add a version check
```
int(importlib.metadata.version('numpy').split('.')[0]) >= 2
```

SamLubelsky committed:
added numpy version checks to determine if copy=False should be tested

also, fixed a little bug in test_matexpr which was
causing the test to fail on master.

SamLubelsky committed:
fixed trailing whitespace

oscarbenjamin commented:
Looks good. Thanks
